include/maths/CDataFrameCategoryEncoder.h

```diff
@@ -85,60 +94,127 @@ class CMakeDataFrameCategoryEncoder;
 //! the number of features we use in total based on the quantity of training data.
 class MATHS_EXPORT CDataFrameCategoryEncoder final {
 public:
-    using TBoolVec = std::vector<bool>;
     using TDoubleVec = std::vector<double>;
-    using TDoubleVecVec = std::vector<TDoubleVec>;
     using TSizeVec = std::vector<std::size_t>;
-    using TSizeVecVec = std::vector<TSizeVec>;
     using TRowRef = core::CDataFrame::TRowRef;
 
+    //! \brief Base type of category encodings.
+    class MATHS_EXPORT CEncoding {
+    public:
+        CEncoding(std::size_t inputColumnIndex, double mic);
+        virtual ~CEncoding() = default;
+        virtual EEncoding type() const = 0;
+        virtual double encode(double value) const = 0;
+        virtual std::uint64_t checksum() const = 0;
+        virtual bool isBinary() const = 0;
+
+        std::size_t inputColumnIndex() const;
+        double encode(const TRowRef& row) const;
+        double mic() const;
+
+        //! Persist by passing information to \p inserter.
+        void acceptPersistInserter(core::CStatePersistInserter& inserter) const;
+
+        //! Populate the object from serialized data.
+        bool acceptRestoreTraverser(core::CStateRestoreTraverser& traverser);
+
+        //! return encoding type as string
+        virtual std::string typeString() const = 0;
```

Comment:
nit: I find keeping public virtual methods together in block aids readability.

Reply:
Done